cachi2/core/models/sbom.py

```diff
@@ -93,3 +98,207 @@ class Sbom(pydantic.BaseModel):
+        return SPDXSbom(
+            packages=packages,
+            creationInfo=SPDXCreationInfo(
+                creators=[f"{tool.vendor} {tool.name}" for tool in self.metadata.tools]
+            ),
+        )
+
+
+class SPDXPackageExternalRef(pydantic.BaseModel):
+    """SPDX Package External Reference.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/#721-external-reference-field
+    """
+
+    referenceCategory: str = pydantic.Field(
+        pattern="^SECURITY|PACKAGE-MANAGER|PERSISTENT-ID|OTHER$"
+    )
+    referenceLocator: str
+    referenceType: str
+
+    _validators: ClassVar[dict[str, Union[Callable, dict[str, Callable]]]] = {
+        "SECURITY": {
+            "cpe22": re.compile(r"[c][pP][eE]:/[AHOaho]?(:[A-Za-z0-9\._\-~%]*){0,6}").match,
+            "cpe23": re.compile(
+                r"""cpe:2\.3:[aho\*\-](:(((\?*|\*?)([a-zA-Z0-9\-\._]|(\\[\\\*\?!"#$$%&'\(\)\+,/:;<=>@\[\]\^`\{\|}~]))+(\?*|\*?))|[\*\-])){5}(:(([a-zA-Z]{2,3}(-([a-zA-Z]{2}|[0-9]{3}))?)|[\*\-]))(:(((\?*|\*?)([a-zA-Z0-9\-\._]|(\\[\\\*\?!"#$$%&'\(\)\+,/:;<=>@\[\]\^`\{\|}~]))+(\?*|\*?))|[\*\-])){4}"""
+            ).match,
+            "advisory": urlparse,
+            "fix": urlparse,
+            "url": urlparse,
+            "swid": urlparse,
+        },
+        "PACKAGE-MANAGER": {
+            "maven-central": re.compile(r"^[^:]+:[^:]+(:[^:]+)?$").match,
+            "npm": re.compile(r"^[^@]+@[^@]+$").match,
+            "nuget": re.compile(r"^[^\/]+\/[^\/]+$").match,
+            "bower": re.compile(r"^[^#]+#[^#]+$").match,
+            "purl": urlparse,
+        },
+        "PERSISTENT-ID": {
+            "swh": re.compile(r"^swh:1:(cnt|dir|rev|rel|snp):[a-f0-9]{40}$").match,
+            "gitoid": re.compile(
+                r"^gitoid:(blob|tree|commit|tag):(sha1|sha256):([a-f0-9]{40}|[a-f0-9]{64})$"
+            ).match,
+        },
+        "OTHER": re.compile(r"[^ ]*").match,
+    }
+
+    @pydantic.model_validator(mode="after")
+    def _model_validator(self: Self) -> Self:
+        if self.referenceCategory == "SECURITY":
+            if self.referenceType not in ["cpe22", "cpe23", "advisory", "fix", "url", "swid"]:
+                raise ValueError("Invalid reference type for SECURITY category")
+        elif self.referenceCategory == "PACKAGE-MANAGER":
+            if self.referenceType not in ["maven-central", "npm", "nuget", "bower", "purl"]:
+                raise ValueError("Invalid reference type for PACKAGE-MANAGER category")
+        elif self.referenceCategory == "PERSISTENT-ID":
+            if self.referenceType not in ["swh", "gitoid"]:
+                raise ValueError("Invalid reference type for PERSISTENT-ID category")
+        if self.referenceCategory == "OTHER" and not cast(
+            Callable, self._validators[self.referenceCategory]
+        )(self.referenceLocator):
+            raise ValueError("Invalid reference locator")
+
+        return self
```

Comment:
What do you think about something along the following lines:
```suggestion
class BaseReference(BaseModel):                                                                                                                           
   def __hash__(self):                                                                                                                                   
        return hash((self.ref_cat, self.ref_type, self.ref_locator))                                                                                         
                                                                                                                                                          
class SecurityReference(BaseReference):                                                                                                                   
    ref_cat: Literal["SEC"]                                                                                                                               
    ref_type: Literal["cpe22", "cpe23"]                                                                                                                   
    ref_locator: str                                                                                                                                         
                                                                                                                                                          
class PMReference(BaseReference):                                                                                                                         
    ref_cat: Literal["PM"]                                                                                                                                
    ref_type: Literal["npm", "maven"]                                                                                                                     
    ref_locator: str   
 
ExtRef = Annotated[Union[SecurityReference, PMReference], Field(discriminator='ref_cat')]

class SPDXPackage(BaseModel):                                                                                                                                      
    ers: list[ExtRef]
    ....
```

This is a PoC so I have cut down the number of literals and also names too, but it should give an idea of what I mean. This approach could be taken further with regexps for reference locators.

Reply:
Yeah, that could work. I didn't know about this pydantic feature